homeassistant/components/upnp/config_flow.py

```diff
@@ -1,10 +1,183 @@
 """Config flow for UPNP."""
+from typing import Mapping, Optional
+from urllib.parse import urlparse
+
+import voluptuous as vol
+
 from homeassistant import config_entries
-from homeassistant.helpers import config_entry_flow
+from homeassistant.components import ssdp
 
-from .const import DOMAIN
+from .const import (  # pylint: disable=unused-import
+    CONFIG_ENTRY_ST,
+    CONFIG_ENTRY_UDN,
+    DISCOVERY_LOCATION,
+    DISCOVERY_NAME,
+    DISCOVERY_ST,
+    DISCOVERY_UDN,
+    DISCOVERY_USN,
+    DOMAIN,
+    LOGGER as _LOGGER,
+)
 from .device import Device
 
-config_entry_flow.register_discovery_flow(
-    DOMAIN, "UPnP/IGD", Device.async_discover, config_entries.CONN_CLASS_LOCAL_POLL
-)
+
+class UpnpFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a UPnP/IGD config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    # Paths:
+    # - ssdp(discovery_info) --> ssdp_confirm(None) --> ssdp_confirm({}) --> create_entry()
+    # - user(None): scan --> user({...}) --> create_entry()
+    # - import(None) --> create_entry()
+
+    def __init__(self):
+        """Initialize the UPnP/IGD config flow."""
+        self._discoveries: Mapping = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow start."""
+        _LOGGER.debug("async_step_user: user_input: %s", user_input)
+        # This uses DISCOVERY_USN as the unique_id for user input.
+
+        if user_input is not None:
+            # Ensure wanted device was discovered.
+            matching_discoveries = [
+                discovery
+                for discovery in self._discoveries
+                if discovery[DISCOVERY_USN] == user_input["unique_id"]
+            ]
+            if not matching_discoveries:
+                return self.async_abort(reason="no_devices_discovered")
+
+            discovery = matching_discoveries[0]
+            await self.async_set_unique_id(
+                discovery[DISCOVERY_USN], raise_on_progress=False
+            )
+            return await self._async_create_entry_from_data(discovery)
+
+        # Discover devices.
+        discoveries = await Device.async_discover(self.hass)
+
+        # Filter discoveries which are already configured.
+        current_unique_ids = {
+            entry.unique_id for entry in self._async_current_entries()
+        }
+        self._discoveries = [
+            discovery
+            for discovery in discoveries
+            if discovery[DISCOVERY_USN] not in current_unique_ids
+        ]
+
+        # Ensure anything to add.
+        if not self._discoveries:
+            return self.async_abort(reason="no_devices_found")
+
+        data_schema = vol.Schema(
+            {
+                vol.Required("unique_id"): vol.In(
```

Comment:
Why don't we call this "host" ?

Reply:
I've renamed it to `usn`, as that is what it is. Host itself is not unique enough for some devices, as per #33827.